drivers/flash/flash_cadence_nand.c

```diff
@@ -0,0 +1,292 @@
+	DEVICE_MMIO_NAMED_ROM(sdma);
+#if CDNS_NAND_RESET_SUPPORT
+	/* Reset controller device configuration for Nand*/
+	const struct reset_dt_spec reset;
+	/* Reset controller device configuration for Combo Phy*/
+	const struct reset_dt_spec combo_phy_reset;
+#endif
+#if CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	void (*irq_config)(void);
+#endif
+};
+
+static const struct flash_parameters flash_cdns_parameters = {.write_block_size = FLASH_WRITE_SIZE,
+							      .erase_value = 0xFF};
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+
+struct flash_pages_layout flash_cdns_pages_layout;
+
+void flash_cdns_page_layout(const struct device *nand_dev, const struct flash_pages_layout **layout,
+			    size_t *layout_size)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+
+	flash_cdns_pages_layout.pages_count =
+		(nand_param->npages_per_block * nand_param->nblocks_per_lun * nand_param->nluns);
+	flash_cdns_pages_layout.pages_size = nand_param->page_size;
+	*layout = &flash_cdns_pages_layout;
+	*layout_size = 1;
+}
+
+#endif
+
+static int flash_cdns_nand_erase(const struct device *nand_dev, off_t offset, size_t len)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if (len == 0) {
+		LOG_ERR("Invalid input parameter for Nand Flash Erase!");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&nand_data->nand_mutex, K_FOREVER);
+
+	ret = cdns_nand_erase(nand_param, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
+		LOG_ERR("Cadence Nand Flash Erase Failed!");
+		return ret;
+	}
+
+	k_mutex_unlock(&nand_data->nand_mutex);
+
+	return 0;
+}
+
+static int flash_cdns_nand_write(const struct device *nand_dev, off_t offset, const void *data,
+				 size_t len)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if ((data == NULL) || (len == 0)) {
+		LOG_ERR("Invalid input parameter for Nand Flash Write!");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&nand_data->nand_mutex, K_FOREVER);
+
+	ret = cdns_nand_write(nand_param, data, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
+		LOG_ERR("Cadence Nand Flash Write Failed!");
+		return ret;
+	}
+
+	k_mutex_unlock(&nand_data->nand_mutex);
+
+	return 0;
+}
+
+static int flash_cdns_nand_read(const struct device *nand_dev, off_t offset, void *data, size_t len)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if ((data == NULL) || (len == 0)) {
+		LOG_ERR("Invalid input parameter for Nand Flash Read!");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&nand_data->nand_mutex, K_FOREVER);
+
+	ret = cdns_nand_read(nand_param, data, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
```

Comment:
same as in erase.

Reply:
@de-nordic as per review comments changes updated